# Ensembl - Alignment course

We will use for this course the genomic data generated for [_Oryza sativa_ Japonica](https://plants.ensembl.org/Oryza_sativa/Info/Index) (rice) 

## Installation of dependencies
To install all the necessary software required in this course we will mainly use [Conda](https://docs.conda.io/en/latest/), which is a package and environment management system that is very convenient for installing all of the dependencies we will use.

1. First, create a new conda environment named `masters_jan2020`

        conda create --name masters_jan2020
2. Check that the new environment exists by doing

        conda env list
3. Activate envirnoment

        conda activate masters_jan2020
4. Install the needed software

        conda install -c bioconda bwa
        conda install -c bioconda samtools
        conda install fastqc

## 1. Material used in this course

* [3000 genomes project dataset](http://iric.irri.org/resources/3000-genomes-project), which is an international effort to resequence a core collection of >3000 rice accessions from 92 countries.

## 2. Preparing the reference genome

* Download the latest rice genome assembly from:

[ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/oryza_sativa/dna/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz](ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/oryza_sativa/dna/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz)

This compressed FASTA file contain all sequence regions flagged as toplevel in an Ensembl schema. This includes chromsomes, regions not assembled into chromosomes and N padded haplotype/patch regions

Now, decompress the gzipped FASTA file

        gzip -d Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz

Finally, you need to generate an index file that will be used by different tools in this course

        samtools faidx Oryza_sativa.IRGSP-1.0.dna.toplevel.fa

## 3. Download sequencing data for sample SAMEA2569438 
Download the two FASTQ files containing the forward and reverse reads in a paired-end sequencing experiment from the ENA:

[https://www.ebi.ac.uk/ena/browser/view/SAMEA2569438](https://www.ebi.ac.uk/ena/browser/view/SAMEA2569438)

### 3.1 Quality control of the FASTQ files
For this, we will use [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). In order to use this software with the downloaded FASTQ files run:

    fastqc ERR605441/ERR605441_1.fastq.gz ERR605441/ERR605441_2.fastq.gz 

And then <mark>browser used in VM need to be defined</mark> open with your browser  the `ERR605445_1_fastqc.html` and `ERR605445_2_fastqc.html` reports generated by `FASTQC`
## 4. Alignment

<mark>Reference material can be found [here]</mark>(https://gatk.broadinstitute.org/hc/en-us/articles/360035535912-Data-pre-processing-for-variant-discovery)

In this course we are going to use [BWA](http://bio-bwa.sourceforge.net/) for aligning the short reads to the reference. There are other reading mapping tools, but BWA is one of the most popular and the one we have the most experience with in our group. This tool, requires a preprocessing step consisting on building an index from the reference FASTA sequence that has been downloaded in section 2. In order to build this index do:

    bwa index /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz

Index construction takes a while but the good thing is that this index can be resued for different alignment runs

### 4.1 BWA mem
BWA has different run options. Enter this in your terminal:

    bwa

    You will get something very similar to:

    Program: bwa (alignment via Burrows-Wheeler transformation)
    Version: 0.7.17-r1188
    Contact: Heng Li <lh3@sanger.ac.uk>

    Usage:   bwa <command> [options]

    Command: index         index sequences in the FASTA format
         mem           BWA-MEM algorithm
         fastmap       identify super-maximal exact matches
         pemerge       merge overlapping paired ends (EXPERIMENTAL)
         aln           gapped/ungapped alignment
         samse         generate alignment (single ended)
         sampe         generate alignment (paired ended)
         bwasw         BWA-SW for long queries

         shm           manage indices in shared memory
         fa2pac        convert FASTA to PAC format
         pac2bwt       generate BWT from PAC
         pac2bwtgen    alternative algorithm for generating BWT
         bwtupdate     update .bwt to the new format
         bwt2sa        generate SA from BWT and Occ

    Note: To use BWA, you need to first index the genome with `bwa index'.
      There are three alignment algorithms in BWA: `mem', `bwasw', and
      `aln/samse/sampe'. If you are not sure which to use, try `bwa mem'
      first. Please `man ./bwa.1' for the manual.

It is widely accepted that the alignment for reads having a length: 70 bp to 1Mb is `BWA mem`, as it is more accurate and faster than the other alignment algorithms included with `BWA`

 To run `BWA` enter the following if you want to get an aligment in the SAM format:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa runId_1.fastq.gz runId_2.fastq.gz -o ${runId}.sam

If you want to obtain an alignment file in the `BAM` format enter the following:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa runId_1.fastq.gz runId_2.fastq.gz |samtools view -b -o ${runId}.bam

Finally, if you want to generate an alignment in the CRAM format do the following:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa runId_1.fastq.gz runId_2.fastq.gz |samtools view -C -T /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa > ${runId}.cram

### 4.2 The SAM alignment format
The Sequence Alignment Format (SAM) is a text-based format (check the Wikipedia [entry](https://en.wikipedia.org/wiki/SAM_(file_format))), used for representing the alignment of the short reads in the FASTQ files to the reference sequence.
This format is not compressed and it is preferable to convert it to its binary equivalent (BAM) or to the ultra-compressed equivalent called CRAM to facilitate its handling and storage.

Compare the diferent file sizes for each of the alignment files by listing the directory content:

        $ ls -lh

And you get:

        $ drwxrwxr-x 2 ernesto nucleotide  4.0K Nov  7 19:21 .
        $ drwxrwxr-x 5 ernesto nucleotide  4.0K Nov  7 19:03 ..
        $ -rw-rw-r-- 1 ernesto nucleotide  393M Nov  6 19:44 ERR605441.bam
        $ -rw-rw-r-- 1 ernesto nucleotide  160M Nov  7 19:21 ERR605441.cram
        $ -rw-rw-r-- 1 ernesto nucleotide  1.1G Nov  6 18:24 ERR605441.sam

<mark>CRAM format reference (https://www.internationalgenome.org/faq/what-are-cram-files/)</mark>

Alignment files in this format (format specification is described [here](https://samtools.github.io/hts-specs/SAMv1.pdf)) start with an optional header section followed by the alignment lines. All header lines start with '@' and are used to represent different metadata elements like the reference version and the chromosome sequence ids used in the alignment, the technology used for generating the sequence data, if the alignment file is sorted or unsorted, etc ... The alignment lines are characterized by having 11 mandatory fields:

| Col         | Field       | Type     | Brief description                     |
| ----------- | ----------- | -------- | ------------------------------------- |
| 1           | QNAME       |  String  |  Query template NAME                  |
| 2           | FLAG        |  int     |  bitwise FLAG                         |
| 3           | RNAME       |  String  |  References sequence NAME             |
| 4           | POS         |  int     |  1- based leftmost mapping POSition   |
| 5           | MAPQ        |  int     |  Mapping quality                      |
| 6           | CIGAR       |  String  |  CIGAR string                         |
| 7           | RNEXT       |  String  |  Ref. name of the mate/next read      |
| 8           | PNEXT       |  int     |  Position of the mate/next read       |
| 9           | TLEN        |  int     |  observed template length             |
| 10          | SEQ         |  String  |  segment sequence                     |
| 11          | QUAL        |  String  |  ASCII of Phred-scaled base QUALity+33|

### 4.3 Samtools 

[Samtools](http://www.htslib.org/doc/samtools.html) is a command line tool used to interact and manipulate the SAM-related alignment files. These are some of the set of commands that are more relevant for this course:

        $ samtools view ERR605441.bam # print the alignment to stdout 

        $ samtools view -H ERR605441.bam # print the header section

And if you want to get the alignments on a particular genomic region:

        $ samtools view ERR605441.bam 1:10000000-10000010
        [main_samview] random alignment retrieval only works for indexed BAM or CRAM files.

This does not work because you need to index the `.bam` file first:


 * Some basic stats on the alignment

        $ samtools stats ERR605441.bam |less

And you get:

        SN      raw total sequences:    4293094
        SN      filtered sequences:     0
        SN      sequences:      4293094
        SN      is sorted:      0
        SN      1st fragments:  2146547
        SN      last fragments: 2146547
        SN      reads mapped:   4094201
        SN      reads mapped and paired:        4073468 # paired-end technology bit set + both mates mapped
        SN      reads unmapped: 198893
        SN      reads properly paired:  3822534 # proper-pair bit set
        SN      reads paired:   4293094 # paired-end technology bit set
        SN      reads duplicated:       0       # PCR or optical duplicate bit set
        SN      reads MQ0:      719477  # mapped and MQ=0
        SN      reads QC failed:        0
        SN      non-primary alignments: 0
        SN      total length:   356326802       # ignores clipping
        SN      total first fragment length:    178163401       # ignores clipping
        SN      total last fragment length:     178163401       # ignores clipping
        SN      bases mapped:   339818683       # ignores clipping
        SN      bases mapped (cigar):   337723141       # more accurate
        SN      bases trimmed:  0
        SN      bases duplicated:       0
        SN      mismatches:     2477885 # from NM fields
        SN      error rate:     7.337030e-03    # mismatches / bases mapped (cigar)
        SN      average length: 83
        SN      average first fragment length:  83
        SN      average last fragment length:   83
        SN      maximum length: 83
        SN      maximum first fragment length:  83
        SN      maximum last fragment length:   83
        SN      average quality:        36.7
        SN      insert size average:    451.2
        SN      insert size standard deviation: 36.9
        SN      inward oriented pairs:  1970730
        SN      outward oriented pairs: 3846
        SN      pairs with other orientation:   4771
        SN      pairs on different chromosomes: 58807
        SN      percentage of properly paired reads (%):        89.0
