# Ensembl - Alignment course

We will use for this course the genomic data generated for [_Oryza sativa_ Japonica](https://plants.ensembl.org/Oryza_sativa/Info/Index) (rice) 

## Installation of dependencies
To install all the necessary software required in this course we will mainly use [Conda](https://docs.conda.io/en/latest/), which is a package and environment management system that is very convenient for installing all of the dependencies we will use.

1. First, create a new conda environment named `masters_jan2020`

        conda create --name masters_jan2020
2. Check that the new environment exists by doing

        conda env list
3. Activate envirnoment

        conda activate masters_jan2020
4. Install the needed software

        conda install -c bioconda bwa
        conda install -c bioconda samtools
        conda install fastqc

## 1. Material used in this course

* [3000 genomes project dataset](http://iric.irri.org/resources/3000-genomes-project), which is an international effort to resequence a core collection of >3000 rice accessions from 92 countries.

## 2. Preparing the reference genome

* Download the latest rice genome assembly from:

[ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/oryza_sativa/dna/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz](ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/oryza_sativa/dna/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz)

This compressed FASTA file contain all sequence regions flagged as toplevel in an Ensembl schema. This includes chromsomes, regions not assembled into chromosomes and N padded haplotype/patch regions

Now, decompress the gzipped FASTA file

        gzip -d Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz

Finally, you need to generate an index file that will be used by different tools in this course

        samtools faidx Oryza_sativa.IRGSP-1.0.dna.toplevel.fa

## 3. Download sequencing data for sample SAMEA2569438 
Download the two FASTQ files containing the forward and reverse reads in a paired-end sequencing experiment from the ENA:

[https://www.ebi.ac.uk/ena/browser/view/SAMEA2569438](https://www.ebi.ac.uk/ena/browser/view/SAMEA2569438)

### 3.1 Quality control of the FASTQ files
For this, we will use [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). In order to use this software with the downloaded FASTQ files run:

    fastqc ERR605441/ERR605441_1.fastq.gz ERR605441/ERR605441_2.fastq.gz 

And then <mark>browser used in VM need to be defined</mark> open with your browser  the `ERR605445_1_fastqc.html` and `ERR605445_2_fastqc.html` reports generated by `FASTQC`
## 4. Alignment

<mark>Reference material can be found [here]</mark>(https://gatk.broadinstitute.org/hc/en-us/articles/360035535912-Data-pre-processing-for-variant-discovery)

In this course we are going to use [BWA](http://bio-bwa.sourceforge.net/) for aligning the short reads to the reference. There are other reading mapping tools, but BWA is one of the most popular and the one we have the most experience with in our group. This tool, requires a preprocessing step consisting on building an index from the reference FASTA sequence that has been downloaded in section 2. In order to build this index do:

    bwa index /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz

Index construction takes a while but the good thing is that this index can be resued for different alignment runs

### 4.1 BWA mem
BWA has different run options. Enter this in your terminal:

    bwa

    You will get something very similar to:

    Program: bwa (alignment via Burrows-Wheeler transformation)
    Version: 0.7.17-r1188
    Contact: Heng Li <lh3@sanger.ac.uk>

    Usage:   bwa <command> [options]

    Command: index         index sequences in the FASTA format
         mem           BWA-MEM algorithm
         fastmap       identify super-maximal exact matches
         pemerge       merge overlapping paired ends (EXPERIMENTAL)
         aln           gapped/ungapped alignment
         samse         generate alignment (single ended)
         sampe         generate alignment (paired ended)
         bwasw         BWA-SW for long queries

         shm           manage indices in shared memory
         fa2pac        convert FASTA to PAC format
         pac2bwt       generate BWT from PAC
         pac2bwtgen    alternative algorithm for generating BWT
         bwtupdate     update .bwt to the new format
         bwt2sa        generate SA from BWT and Occ

    Note: To use BWA, you need to first index the genome with `bwa index'.
      There are three alignment algorithms in BWA: `mem', `bwasw', and
      `aln/samse/sampe'. If you are not sure which to use, try `bwa mem'
      first. Please `man ./bwa.1' for the manual.

It is widely accepted that the alignment for reads having a length: 70 bp to 1Mb is `BWA mem`, as it is more accurate and faster than the other alignment algorithms included with `BWA`

 To run `BWA` enter the following if you want to get an aligment in the SAM format:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz runId_1.fastq.gz runId_2.fastq.gz |samtools view -b -o ${runId}.bam