# IGSR - Alignment course

We will use for this course the genomic data generated for [_Oryza sativa_ Japonica](https://plants.ensembl.org/Oryza_sativa/Info/Index) (rice) 

## Installation of dependencies
To install all the necessary software required in this course we will mainly use [Conda](https://docs.conda.io/en/latest/), which is a package and environment management system that is very convenient for installing all of the dependencies we will use.

1. First, create a new conda environment named `masters_jan2020`

        conda create --name masters_jan2020
2. Check that the new environment exists by doing

        conda env list
3. Activate envirnoment

        conda activate masters_jan2020
4. Install the necessary software

        conda install -c bioconda bwa
        conda install -c bioconda samtools
        conda install fastqc
        conda install picard
        conda install -c bioconda igv

## 1. Material used in this course

* [3000 genomes project dataset](http://iric.irri.org/resources/3000-genomes-project), which is an international effort to resequence a core collection of >3000 rice accessions from 92 countries.

## 2. Preparing the reference genome

* Download the latest rice genome assembly from:

[ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/oryza_sativa/dna/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz](ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/oryza_sativa/dna/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz)

This compressed FASTA file contain all sequence regions flagged as toplevel in an Ensembl schema. This includes chromsomes, regions not assembled into chromosomes and N padded haplotype/patch regions

Now, decompress the gzipped FASTA file

        gzip -d Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz

Finally, you need to generate an index file that will be used by different tools in this course

        samtools faidx Oryza_sativa.IRGSP-1.0.dna.toplevel.fa

## 3. Download sequencing data for sample SAMEA2569438 
Download the two FASTQ files containing the forward and reverse reads in a paired-end sequencing experiment from:

<mark>[https://www.ebi.ac.uk/ena/browser/view/SAMEA2569438](https://www.ebi.ac.uk/ena/browser/view/SAMEA2569438)

### 3.1 Quality control of the FASTQ files
For this, we will use [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). In order to use this software with the downloaded FASTQ files run:

    fastqc SAMEA2569438.chr10_1.fastq.gz SAMEA2569438.chr10_2.fastq.gz 

And then <mark>browser used in VM need to be defined</mark> open with your browser  the `SAMEA2569438.chr10_1.html` and `SAMEA2569438.chr10_2.html` reports generated by `FASTQC`
## 4. Alignment

<mark>Reference material can be found [here]</mark>(https://gatk.broadinstitute.org/hc/en-us/articles/360035535912-Data-pre-processing-for-variant-discovery)

In this course we are going to use [BWA](http://bio-bwa.sourceforge.net/) for aligning the short reads to the reference. There are other reading mapping tools, but BWA is one of the most popular and the one we have the most experience with in our group. This tool, requires a preprocessing step consisting on building an index from the reference FASTA sequence that has been downloaded in section 2. In order to build this index do:

    bwa index /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa.gz

Index construction takes a while but the good thing is that this index can be resued for different alignment runs

### 4.1 BWA mem
BWA has different run options. Enter this in your terminal:

    bwa

    You will get something very similar to:

    Program: bwa (alignment via Burrows-Wheeler transformation)
    Version: 0.7.17-r1188
    Contact: Heng Li <lh3@sanger.ac.uk>

    Usage:   bwa <command> [options]

    Command: index         index sequences in the FASTA format
         mem           BWA-MEM algorithm
         fastmap       identify super-maximal exact matches
         pemerge       merge overlapping paired ends (EXPERIMENTAL)
         aln           gapped/ungapped alignment
         samse         generate alignment (single ended)
         sampe         generate alignment (paired ended)
         bwasw         BWA-SW for long queries

         shm           manage indices in shared memory
         fa2pac        convert FASTA to PAC format
         pac2bwt       generate BWT from PAC
         pac2bwtgen    alternative algorithm for generating BWT
         bwtupdate     update .bwt to the new format
         bwt2sa        generate SA from BWT and Occ

    Note: To use BWA, you need to first index the genome with `bwa index'.
      There are three alignment algorithms in BWA: `mem', `bwasw', and
      `aln/samse/sampe'. If you are not sure which to use, try `bwa mem'
      first. Please `man ./bwa.1' for the manual.

It is widely accepted that the alignment for reads having a length: 70 bp to 1Mb is `BWA mem`, as it is more accurate and faster than the other alignment algorithms included with `BWA`

 To run `BWA` enter the following if you want to get an aligment in the SAM format:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa runId_1.fastq.gz runId_2.fastq.gz -o ${runId}.sam

If you want to obtain an alignment file in the `BAM` format enter the following:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa runId_1.fastq.gz runId_2.fastq.gz |samtools view -b -o ${runId}.bam

Finally, if you want to generate an alignment in the CRAM format do the following:

        bwa mem /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa runId_1.fastq.gz runId_2.fastq.gz |samtools view -C -T /path/to/Oryza_sativa.IRGSP-1.0.dna.toplevel.fa > ${runId}.cram

You can check that `BWA` is running by using the `top` Unix command. For this, first check what is your username in your machine:

        $ whoami
        ernesto # this will be different depending on your system
Now, you can use the `top` commamd to obtain a summary of the processes that user `ernesto` is running in the system:

        top -u ernesto
You should see something similar to what I got in my system:
        
        top - 13:01:44 up 67 days,  9:27,  0 users,  load average: 11.18, 11.14, 11.22
        Tasks: 995 total,   6 running, 989 sleeping,   0 stopped,   0 zombie
        %Cpu(s): 13.8 us,  0.4 sy,  0.0 ni, 85.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
        KiB Mem : 26372707+total, 14490720+free, 68764784 used, 50055096 buff/cache
        KiB Swap:  8191996 total,  7146680 free,  1045316 used. 19283916+avail Mem

        PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
        164759 ernesto   20   0  936492 778536    936 S 100.3  0.3   0:11.97 bwa
        164875 ernesto   20   0  171116   3380   1628 R   0.3  0.0   0:00.07 top
        101804 ernesto   20   0   46216   3152   1792 S   0.0  0.0   4:58.73 res
        101839 ernesto   20   0  113140   1440   1200 S   0.0  0.0   0:00.00 1605100319.8865
        101841 ernesto   20   0  129504   3952   1828 S   0.0  0.0   0:02.51 bash
        164760 ernesto   20   0   87720   2676   1872 S   0.0  0.0   0:00.00 samtools

### 4.2 The SAM alignment format
The Sequence Alignment Format (SAM) is a text-based format (check the Wikipedia [entry](https://en.wikipedia.org/wiki/SAM_(file_format))), used for representing the alignment of the short reads in the FASTQ files to the reference sequence.
This format is not compressed and it is preferable to convert it to its binary equivalent (BAM) or to the ultra-compressed equivalent called CRAM to facilitate its handling and storage.

Compare the diferent file sizes for each of the alignment files by listing the directory content:

        $ ls -lh

And you get:

        drwxrwxr-x 2 ernesto nucleotide 4.0K Nov 11 17:22 .
        drwxrwxr-x 5 ernesto nucleotide 4.0K Nov 11 17:22 ..
        -rw-rw-r-- 1 ernesto nucleotide 108M Nov 11 13:25 SAMEA2569438.chr10.bam
        -rw-rw-r-- 1 ernesto nucleotide  42M Nov 11 17:14 SAMEA2569438.chr10.cram
        -rw-rw-r-- 1 ernesto nucleotide 308M Nov 11 17:05 SAMEA2569438.chr10.sam

<mark>CRAM format reference (https://www.internationalgenome.org/faq/what-are-cram-files/)</mark>

Alignment files in this format (format specification is described [here](https://samtools.github.io/hts-specs/SAMv1.pdf)) start with an optional header section followed by the alignment lines. All header lines start with '@' and are used to represent different metadata elements like the reference version and the chromosome sequence ids used in the alignment, the technology used for generating the sequence data, if the alignment file is sorted or unsorted, etc ... The alignment lines are characterized by having 11 mandatory fields:

| Col         | Field       | Type     | Brief description                     |
| ----------- | ----------- | -------- | ------------------------------------- |
| 1           | QNAME       |  String  |  Query template NAME                  |
| 2           | FLAG        |  int     |  bitwise FLAG                         |
| 3           | RNAME       |  String  |  References sequence NAME             |
| 4           | POS         |  int     |  1- based leftmost mapping POSition   |
| 5           | MAPQ        |  int     |  Mapping quality                      |
| 6           | CIGAR       |  String  |  CIGAR string                         |
| 7           | RNEXT       |  String  |  Ref. name of the mate/next read      |
| 8           | PNEXT       |  int     |  Position of the mate/next read       |
| 9           | TLEN        |  int     |  observed template length             |
| 10          | SEQ         |  String  |  segment sequence                     |
| 11          | QUAL        |  String  |  ASCII of Phred-scaled base QUALity+33|

### 4.3 Samtools 

[Samtools](http://www.htslib.org/doc/samtools.html) is a command line tool used to interact and manipulate the SAM-related alignment files. These are some of the set of commands that are more relevant for this course:

        $ samtools view SAMEA2569438.chr10.bam # print the alignment to stdout 

        $ samtools view -H SAMEA2569438.chr10.bam # print the header section

And if you want to get the alignments on a particular genomic region:

        $ samtools view SAMEA2569438.chr10.bam 10:10000000-10001000
        [main_samview] random alignment retrieval only works for indexed BAM or CRAM files.

This does not work because you need to sort and index the `.bam` file first:

        $ samtools sort SAMEA2569438.chr10.bam -o SAMEA2569438.chr10.sorted.bam -O BAM
        $ samtools index SAMEA2569438.chr10.sorted.bam # index
        $ samtools view SAMEA2569438.chr10.sorted.bam 10:10000000-10001000 # repeat the command

 * Some basic stats on the alignment

        $ samtools flagstat SAMEA2569438.chr10.bam

And you get:

        1174129 + 0 in total (QC-passed reads + QC-failed reads)
        0 + 0 secondary
        759 + 0 supplementary
        0 + 0 duplicates
        1167569 + 0 mapped (99.44% : N/A)
        1173370 + 0 paired in sequencing
        586685 + 0 read1
        586685 + 0 read2
        1121364 + 0 properly paired (95.57% : N/A)
        1160250 + 0 with itself and mate mapped
        6560 + 0 singletons (0.56% : N/A)
        0 + 0 with mate mapped to a different chr
        0 + 0 with mate mapped to a different chr (mapQ>=5)

### Alignment post-processing
The alignment file in the BAM format needs a series of post-processing steps that are required for variant discovery. The different steps that are shown here will produce an analysis-ready BAM file that can be used in the following module of this course

##### **Adding metadata to the alignment file**
BWA generates a `BAM` file without any metadata on the sequencing experimental design that has been used, this is why we need to manually add this metadata so it can by used by the variant calling analysis that is described in the variant calling section of this course. For this, we are going to use [Picard AddOrReplaceReadGroups](https://gatk.broadinstitute.org/hc/en-us/articles/360037226472-AddOrReplaceReadGroups-Picard-) in the following way:

        picard AddOrReplaceReadGroups I=SAMEA2569438.chr10.bam RGSM=SAMEA2569438 RGLB=SAMEA2569438 RGPL=ILLUMINA O=SAMEA2569438.chr10.reheaded.bam RGPU=SAMEA2569438

This command will add information on the sample id that has been sequenced, what sequencing platform has been used and also information on the sequencing library id. You can check the SAM header for the new BAM with metadata information by doing:

        samtools view -H SAMEA2569438.chr10.reheaded.bam

And you will see the added metadata in the `@RG` line:

        @HD     VN:1.6  SO:unsorted
        @SQ     SN:10   LN:23207287
        @RG     ID:1    LB:SAMEA2569438 PL:ILLUMINA     SM:SAMEA2569438 PU:SAMEA2569438
        @PG     ID:bwa  PN:bwa  VN:0.7.17-r1188 CL:bwa mem Oryza_sativa.IRGSP-1.0.dna.toplevel.chr10.fa SAMEA2569438.chr10_1.fastq.gz SAMEA2569438.chr10_2.fastq.gz

##### **Sort the alignment file**
Most of the tools used for post-processing assume that the `BAM` file is coordinate sorted. We are going to use `samtools sort` for sorting the alignment file:

        samtools sort SAMEA2569438.chr10.reheaded.bam -O BAM -o SAMEA2569438.chr10.reheaded.sorted.bam

##### **MarkDuplicates** 
The alignment file can contain reads that are duplicates. These reads are originated in the PCR amplification step during the sample preparation that might produce identical reads coming from the same DNA fragment. These duplicate reads need to be identified and marked so they can be correctly handled by the variant calling tool. There are multiple tools to detect these duplicates, in this course we will use [Picard MarkDuplicates](https://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates), which is one of the most reliable ones.

        picard MarkDuplicates I=SAMEA2569438.chr10.reheaded.sorted.bam O=SAMEA2569438.chr10.reheaded.sorted.mark_duplicates.bam M=SAMEA2569438.chr10.metrics.txt

With use option `M` so `MarkDuplicates` generates a text file with metrics on the number of reads duplicates. This file is named ` SAMEA2569438.chr10.metrics.txt` in this case. Let's open this metrics file:

        less SAMEA2569438.chr10.metrics.txt
The first part of the file is the most relevant for us:

      ## METRICS CLASS        picard.sam.DuplicationMetrics
      LIBRARY UNPAIRED_READS_EXAMINED READ_PAIRS_EXAMINED     SECONDARY_OR_SUPPLEMENTARY_RDS  UNMAPPED_READS  UNPAIRED_READ_DUPLICATES        READ_PAIR_DUPLICATES    READ_PAIR_OPTICAL_DUPLICATES    PERCENT_DUPLICATION     ESTIMATED_LIBRARY_SIZE
      SAMEA2569438    6560    580125  759     6560    524     19812   0       0.034408        8298967
        ...

We can the reads that are duplicates using samtools view in combination with the bitwise FLAG value in the second column that retrieves the PCR duplicates

        samtools view -f 1024 SAMEA2569438.chr10.reheaded.sorted.mark_duplicates.bam |less

### **Viewing the aligned reads using IGV**
The Integrative Genomics Viewer [IVG](http://software.broadinstitute.org/software/igv/) is a useful interactive tool that can be used to explore visually your genomic data. We are going to use it here to display the alignments we have generated. In this example we will fetch the alignments for a specific region in chromosome 10.

For this, we need first to index the `BAM` file:

        samtools index SAMEA2569438.chr10.reheaded.sorted.mark_duplicates.bam

 Then we extract all the alignments for the `10:10000000-11000000` region that are correct, which means that both members of the read pair are correctly mapped. For this, we use the SAM flag = 2 and `samtools view`:

        samtools view -f 2 ../SAMEA2569438.chr10.reheaded.sorted.mark_duplicates.bam 10:10000000-11000000 -b -o SAMEA2569438.chr10.reheaded.sorted.mark_duplicates.mini.bam

And now we need to create an index for the new `BAM`, as IGV needs it to quickly retrieve the aligments to display:

        samtools index SAMEA2569438.chr10.reheaded.sorted.mark_duplicates.mini.bam

Now, open `IGV` by going to your terminal and entering:

        igv

You will need to load in `IGV` the FASTA file containing the chromosome 10 sequence for rice, as this sequence is not included by default in `IGV`:

![load_genome_igv](https://www.ebi.ac.uk/~ernesto/IGSR/masters_IAMZ_jan2020/load_genome.png)

Look for you file and open it.

Now, load the `GTF` file containing the rice gene annotations for chromosome 10:

![load_annotation_igv](https://www.ebi.ac.uk/~ernesto/IGSR/masters_IAMZ_jan2020/load_annotation.png)

![annotation_view_igv](https://www.ebi.ac.uk/~ernesto/IGSR/masters_IAMZ_jan2020/annotation_view.png)

You will see the new track with genes annotated in chromosome 10

Now, load the BAM file with your alignments:
![load_alignments_igv](https://www.ebi.ac.uk/~ernesto/IGSR/masters_IAMZ_jan2020/load_alignments.png)

![alignments_view1_igv](https://www.ebi.ac.uk/~ernesto/IGSR/masters_IAMZ_jan2020/alignments_view1.png)

And you will see two new tracks, one with the alignments and the other with the depth of coverage. The alignments still do not appear, as the region is too large to render the data.

![alignments_view2_igv](https://www.ebi.ac.uk/~ernesto/IGSR/masters_IAMZ_jan2020/alignments_view2.png)

Enter the genomic interval you want to inspect in the blank box or click on the '-' or '+' signs to zoom in/out




